i2c/mcp9808_i2c/mcp9808_i2c.c

```diff
@@ -0,0 +1,161 @@
+/**
+ * Copyright (c) 2020 Raspberry Pi (Trading) Ltd.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include "pico/stdlib.h"
+#include "pico/binary_info.h"
+#include "hardware/i2c.h"
+
+/* Example code to talk to a MCP9808 ±0.5°C Digital temperature Sensor
+   
+   This reads and writes to registers on the board. 
+
+   Connections on Raspberry Pi Pico board, other boards may vary.
+
+   GPIO PICO_DEFAULT_I2C_SDA_PIN (On Pico this is 4 (pin 6)) -> SDA on MCP9808 board
+   GPIO PICO_DEFAULT_I2C_SCK_PIN (On Pico this is 5 (pin 7)) -> SCL on MCP9808 board
+   Vsys (pin 39) -> VDD on MCP9808 board
+   GND (pin 38)  -> GND on MCP9808 board
+
+*/
+//The bus address is determined by the state of pins A0, A1 and A2 on the MCP9808 board
+static uint8_t ADDRESS = 0X18;
+
+const uint8_t WRITE_MODE = 0XFE;
+const uint8_t READ_MODE = 0XFF;
+
+
+//hardware registers
+
+const uint8_t REG_POINTER = 0X00;
+const uint8_t REG_CONFIG = 0X01;
+const uint8_t REG_TEMP_UPPER = 0X02;
+const uint8_t REG_TEMP_LOWER = 0X03;
+const uint8_t REG_TEMP_CRIT = 0X04;
+const uint8_t REG_TEMP_AMB = 0X05;
+const uint8_t REG_MAN_ID = 0X06;
+const uint8_t REG_DEV_ID = 0X07;
+const uint8_t REG_RESOLUTION = 0X08;
+
+
+void mcp9808_check_limits(uint8_t upper_byte) {
+
+    // Check flags and raise alerts accordingly 
+    if ((upper_byte & 0x40) == 0x40){ //TA > TUPPER
+        printf("Temperature is above the upper temperature limit.\n");
+    }
+    if ((upper_byte & 0x20) == 0x20) { //TA < TLOWER
+        printf("Temperature is below the lower temperature limit.\n");
+    }
+    if ((upper_byte & 0x80) == 0x80) { //TA < TCRIT
+        printf("Temperature is above the critical temperature limit.\n");
+    }
+}
+
+float mcp9808_convert_temp(uint8_t buf[]){
+
+    uint16_t upper_byte;
+    uint16_t lower_byte;
+    float temperature;
+
+    upper_byte = buf[0];
+    lower_byte = buf[1];
+
+
+    mcp9808_check_limits(upper_byte);
+
+    //Clear flag bits
+    upper_byte = upper_byte & 0x1F; 
+
+
+    //Check if TA <= 0°C and convert to denary accordingly
+    if ((upper_byte & 0x10) == 0x10) { 
+        upper_byte = upper_byte & 0x0F; 
+        temperature = 256 - (((float)upper_byte * 16) + ((float)lower_byte / 16));
+    }
+    else {
+        temperature = (((float)upper_byte * 16) + ((float)lower_byte / 16));
+  
+    }
+    return temperature;
+}
+
+
+void mcp9808_read_temp(){  
+
+    uint8_t buf[2];
+    float temperature;
+
+    while (1) {
+    // Start reading ambient temperature register for 2 bytes
+    i2c_write_blocking (i2c_default, 0x18 & WRITE_MODE ,&REG_TEMP_AMB,1,true);
+    i2c_read_blocking(i2c_default, 0x18 & READ_MODE, buf, 2, false);
+
+    temperature = mcp9808_convert_temp(buf);

```

Comment:
_Perhaps_ it makes sense to change the signature of `mcp9808_convert_temp(uint8_t buf[])` to `mcp9808_convert_temp(uint8_t upper_byte, uint8_t lower_byte)` and then inside the `while` loop you could do:
```
    mcp9808_check_limits(buf[0] & 0xE0);

    temperature = mcp9808_convert_temp(buf[0] & 0x1F, buf[1]);
```
(makes things more logically separate IMHO)

Reply:
Looks much nicer now :+1: